src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,218 @@
+double KDERules<MetricType, KernelType, TreeType>::
+Score(const size_t /* queryIndex */, TreeType& /* referenceNode */)
+{
+  ++scores;
+  traversalInfo.LastScore() = 0.0;
+  return 0.0;
+}
+
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::Rescore(
+    const size_t /* queryIndex */,
+    TreeType& /* referenceNode */,
+    const double oldScore) const
+{
+  // If it's pruned it continues to be pruned.
+  return oldScore;
+}
+
+//! Double-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline double KDERules<MetricType, KernelType, TreeType>::
+Score(TreeType& queryNode, TreeType& referenceNode)
+{
+  double score;
+  // Calculations are not duplicated.
+  bool newCalculations = true;
+  const double minDistance = queryNode.MinDistance(referenceNode);
+  const double maxKernel =
+    kernel.Evaluate(minDistance);
+  const double minKernel =
+    kernel.Evaluate(queryNode.MaxDistance(referenceNode));
+  const double bound = maxKernel - minKernel;
+
+  if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
+  {
+    if ((traversalInfo.LastQueryNode() != NULL) &&
+        (traversalInfo.LastReferenceNode() != NULL) &&
+        (traversalInfo.LastQueryNode()->Point(0) == queryNode.Point(0)) &&
+        (traversalInfo.LastReferenceNode()->Point(0) == referenceNode.Point(0)))
+    {
+      // Don't duplicate calculations.
+      newCalculations = false;
+      lastQueryIndex = queryNode.Point(0);
+      lastReferenceIndex = referenceNode.Point(0);
+    }
+  }
+
+  if (bound <= (absError + relError * minKernel) / referenceSet.n_cols &&
+      newCalculations)
+  {
+    // Auxiliary variables.
+    double kernelValue;
+    kde::KDEStat& referenceStat = referenceNode.Stat();
+    kde::KDEStat& queryStat = queryNode.Stat();
+
+    // If calculating a center is not required.
+    if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
+    {
+      kernelValue = EvaluateKernel(queryNode.Point(0), referenceNode.Point(0));
+    }
+    // Sadly, we have no choice but to calculate the center.
+    else
+    {
+      // Calculate center for each node if it has not been calculated yet.
+      if (!referenceStat.ValidCentroid())
+      {
+        arma::vec referenceCenter;
+        referenceNode.Center(referenceCenter);
+        referenceStat.SetCentroid(std::move(referenceCenter));
+      }
+      if (!queryStat.ValidCentroid())
+      {
+        arma::vec queryCenter;
+        queryNode.Center(queryCenter);
+        queryStat.SetCentroid(std::move(queryCenter));
+      }
+      // Compute kernel value.
+      kernelValue = EvaluateKernel(queryStat.Centroid(),
+                                   referenceStat.Centroid());
+    }
+
+    // Can be paralellized but we avoid it for now because of a compilation
+    // error in visual C++ compiler.
+    // #pragma omp for
+    for (size_t i = 0; i < queryNode.NumDescendants(); ++i)
```

Comment:
Just a note, the Visual C++ error can be resolved if you use `omp_size_t` in the for loop instead of `size_t`.  (That's a workaround we had to create.)

Reply:
Did you want to handle this before merge?  I'm not sure if it gave much speedup.

Reply:
I tested both implementations but it didn't gave much speedup and mostly it made it slower so I removed it.